### PR TITLE
Track runtime variables in control-flow builders

### LINKED ISSUE #10938: Add support for closing over variables to control-flow builders
The control-flow builders need their logic updating to capture variable declarations and usages, in order to manually build up the set of closure variables and local definitions for control-flow blocks.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This adds support for all the new classical runtime variables through
the control-flow builder interface.  In particular, most usefully it
automatically manages the scoping rules for new declarations and inner
variable accesses, and ensures that its built scopes automatically close
over any variables used within them (including making sure nested scopes
do the same thing).

The builder API is factored out a little into an explicit interface
object, with `QuantumCircuit` getting an explicit implementation of
that.  This is done because the number of separate API methods we would
have needed to pass around / infer was getting overly large, and this
allows us to just use standard virtual dispatch to automatically do the
right thing.

Python doesn't have a way to have an object implement an
interface other than by structural (duck) typing, so to avoid name
leakage and collisions, we instead make `QuantumCircuit`'s
implementation a friend class that handles the inner state on its
behalf.

Not everything control-flow-builder related is factored out into the API
because it didn't seem overly useful to do this, especially when the
overridden behaviour would just have been to throw exceptions.

### Details and comments

Depends on #10963 and #10974

Close #10938

### CODE REVIEW

File: qiskit/circuit/controlflow/builder.py [hunk 1]
```diff
@@ -36,6 +38,125 @@
     import qiskit
 
 
+class CircuitScopeInterface(abc.ABC):
+    """An interface that circuits and builder blocks explicitly fulfil, which contains the primitive
```
Comment (mtreinish) [inline qiskit/circuit/controlflow/builder.py]:
```suggestion
    """An interface that circuits and builder blocks explicitly fulfill, which contains the primitive
```

File: qiskit/circuit/controlflow/builder.py [hunk 2]
```diff
@@ -36,6 +38,125 @@
...
+    methods of circuit construction and object validation.
+
+    This allows core circuit methods to be applied to the currently open builder scope, and allows
+    the builders to hook into all places where circuit resources might be used.  This allows the
+    builders to track the resources being used, without getting in the way of
+    :class:`.QuantumCircuit` doing its own thing.
+
+    This is a Qiskit-internal interface."""
```
Comment (mtreinish) [inline qiskit/circuit/controlflow/builder.py]: Per https://github.com/Qiskit/documentation/pull/366 if we don't publish this in the API document this is implicit.

File: qiskit/circuit/controlflow/builder.py [hunk 3]
```diff
@@ -238,26 +361,22 @@ def __init__(
...
-                callback.
+            parent: The scope interface of the containing scope.
             forbidden_message: If a string is given here, a :exc:`.CircuitError` will be raised on
                 any attempts to append instructions to the scope with this message.  This is used by
                 pseudo scopes where the state machine of the builder scopes has changed into a
                 position where no instructions should be accepted, such as when inside a ``switch``
                 but outside any cases.
         """
-        self.instructions: List[CircuitInstruction] = []
+        self._instructions: List[CircuitInstruction] = []
```
Comment (mtreinish) [inline qiskit/circuit/controlflow/builder.py:370]: This will probably conflict with: #11214 which I think is changing this to `CircuitData`.

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -5931,6 +5900,78 @@ def qubit_stop_time(self, *qubits: Union[Qubit, int]) -> float:
...
+        if (out := self.circuit._vars_capture.get(name)) is not None:
+            return out
+        return self.circuit._vars_input.get(name)
+
+    def use_var(self, var):
+        if self.get_var(var.name) != var:
+            raise CircuitError(f"'{var}' is not present in this circuit")
+
+
+def _validate_expr(api: CircuitScopeInterface, node: expr.Expr) -> expr.Expr:
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: I would have called this `circuit_scope` or something else, `api` seems a bit too broad. But it doesn't matter a great deal at the end of the day.

File: qiskit/circuit/quantumcircuit.py [hunk 5]
```diff
@@ -1419,30 +1384,18 @@ def append(
...
+        broadcast_iter = (
+            operation.broadcast_arguments(expanded_qargs, expanded_cargs)
+            if isinstance(operation, Instruction)
+            else Instruction.broadcast_arguments(operation, expanded_qargs, expanded_cargs)
+        )
+        for qarg, carg in broadcast_iter:
+            self._check_dups(qarg)
+            instruction = CircuitInstruction(operation, qarg, carg)
+            api.append(instruction)
+            instructions._add_ref(api.instructions, len(api.instructions) - 1)
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: FWIW, this is kind of an unrelated change, it's more compact and easier to read. But really the only change was from `appender` -> `api.append` which took me a bit longer to see because of the other refactoring.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM the changes are pretty sensible and straightforward once you wrap your head around the new interface for dealing with scoping. I had a couple of comments inline. The only thing that I found a bit hard to follow was referring to the current scope variable as `api` everywhere. I know what the intent there is, but I'd probably just call it `scope`, `circuit_scope`, or `current_scope` to make it a bit more explicit about what the variable is. But it's probably ok to stick with `api` everywhere I guess since I could follow it. The only really actionable comment is a tiny docs nit.

File: qiskit/circuit/controlflow/builder.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/controlflow/builder.py]: Huh, turns out this is a 🇬🇧 thing - "fulfil" is correct to me.  I didn't know we had this spelling difference.

File: qiskit/circuit/controlflow/builder.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/circuit/controlflow/builder.py:370]: It will conflict, but in a way that's super easy to fix and honestly, #11214 unifies the types of the `instructions` circuit scope object in a much more pleasant way.

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: I can change it to that - it's no trouble.

File: qiskit/circuit/quantumcircuit.py (hunk 5 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: The `broadcast_iter` is the only unrelated change. Originally this was only collapsing `appender` into `api.append` (since we already have the `api` object), but since #10827 the `api.instructions` field is dispatched as well, so it was growing to quite a few dispatched names.

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Changed in 28d106b618.

File: qiskit/circuit/controlflow/builder.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/circuit/controlflow/builder.py]: I removed the redundant line in 34b47dd44a, to avoid giving the impression that the statement has meaning - we want people to know that the stability policy is the reason.

**Review by mtreinish — APPROVED** [review]
This LGTM, thanks for the quick update
